### PR TITLE
Add docs on handling sensitive data.

### PR DESCRIPTION
**Pull request by paddycarver**
Add a document on handling sensitive state, as well as outlining our suggestion not to use PGP keys to encrypt state.

### CODE REVIEW

File: content/source/docs/extend/best-practices/sensitive-state.html.md [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+`terraform apply`, and was meant to provide a bit of security in that scenario.
+With the introduction and use of [enhanced remote
+backends](/docs/backends/types/index.html) and especially the availability of
+[Terraform Cloud](https://app.terraform.io/), there are now a variety of
+backends that will encrypt state at rest and will not store the state in
+cleartext on machines running `terraform apply`. This means the original
+problem the PGP key pattern was intended to solve has a better-supported
+solution, and we're deprecating it in favor of that solution.
+
+The strategies also have drawbacks. They can't be reliably interpolated,
```
Comment (nfagerlund) [inline content/source/docs/extend/best-practices/sensitive-state.html.md]:
```suggestion
Even without comparing it to full state encryption, per-value encryption has major drawbacks. Encrypted values can't be reliably interpolated,
```

That transition seemed a bit odd; here's one possible way you could tweak it.

File: content/source/docs/extend/best-practices/sensitive-state.html.md [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+Both of these approaches are discouraged and will be removed from the
+HashiCorp-supported providers over time. This strategy was tailored to a time
+when Terraform's state had to be stored in cleartext on any machine running
+`terraform apply`, and was meant to provide a bit of security in that scenario.
+With the introduction and use of [enhanced remote
+backends](/docs/backends/types/index.html) and especially the availability of
+[Terraform Cloud](https://app.terraform.io/), there are now a variety of
+backends that will encrypt state at rest and will not store the state in
+cleartext on machines running `terraform apply`. This means the original
+problem the PGP key pattern was intended to solve has a better-supported
```
Comment (nfagerlund) [inline content/source/docs/extend/best-practices/sensitive-state.html.md]:
```suggestion
problem the PGP key pattern was intended to solve has a more complete
```

File: content/source/docs/extend/best-practices/sensitive-state.html.md [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+# Handling Sensitive Values in State
+
+Many organizations use Terraform to manage their entire infrastructure, and
+it's inevitable that [sensitive information will find its way into
+Terraform](/docs/state/sensitive-data.html) in these circumstances. There are a
+couple of recommended approaches for managing sensitive state in Terraform.
+
+## Using the `Sensitive` Flag
+
+When working with a field that is likely to contain sensitive information, it
```
Comment (radeksimko) [inline content/source/docs/extend/best-practices/sensitive-state.html.md]: Nitpick, but I'd reword this as _field that contains information that is likely to be considered as sensitive_. The reason being that I'd not want anyone to use this sentence as an excuse to tuck `Sensitive` to all fields "just in case".

File: content/source/docs/extend/best-practices/sensitive-state.html.md [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+is best to set the
+[`Sensitive`](https://godoc.org/github.com/hashicorp/terraform-plugin-sdk/helper/schema#Schema.Sensitive)
+property on its schema to `true`. This will prevent the field's values from
+showing up in CLI output and in Terraform Cloud. It will not encrypt or obscure
```
Comment (radeksimko) [inline content/source/docs/extend/best-practices/sensitive-state.html.md]: I would make `not encrypt` bold here, to make it stand out.

File: content/source/docs/extend/best-practices/sensitive-state.html.md [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+the value in the state, however.
+
+## Storing Only a Hash
+
+Providers can also protect sensitive information by simply not storing it.
```
Comment (radeksimko) [inline content/source/docs/extend/best-practices/sensitive-state.html.md]: Technically we are storing a hash of it and hash could be considered as sensitive, especially if it's something like MD5 or SHA1.

File: content/source/docs/extend/best-practices/sensitive-state.html.md [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+Storing a hash of the information instead of the value itself is popular when:
+
+* The value is set by the provider/the API, and cannot be set in the config
```
Comment (radeksimko) [inline content/source/docs/extend/best-practices/sensitive-state.html.md]: ... _i.e. the field has `Computed` set to `true`_ (to make it clearer?)

File: content/source/docs/extend/best-practices/sensitive-state.html.md (hunk 2 above)
Comment (radeksimko) [reply content/source/docs/extend/best-practices/sensitive-state.html.md]: I feel like changing this to "more complete" is overselling it a bit. It's easier encryption at rest with good UX, but I think we're still far away from solving the problem of sensitive data completely.

Comment (radeksimko) [reply content/source/docs/extend/best-practices/sensitive-state.html.md]: I feel like changing this to "more complete" is overselling it a bit. It's easier encryption at rest with good UX, but I think we're still far away from solving the problem of sensitive data completely.
